NF: add contentOfDirectory method

Arthur-Milchior opened the pull request:
This method returns an iterator over directory content as File.

Starting at API 26, this uses [Files.newDirectoryStream], ensuring that all
operations are constant time and space. For lower API, it simulates this method
using `listFiles`, whose space and time are linear in the number of file of the directory.

This is a first commit created to factorize more the MoveDirectory operation

david-allison reviewed (changes requested):
`DirectoryStream` is a closeable which is never closed.

Returning iterators is typically a risky operation from an architectural perspective 

It's very tricky to ensure that `close` is called - `finalize()` was so bad that it was deprecated in Java 9.

I haven't looked into Kotlin's `sequence`, but this may be a better option.

It would be better to implement a "process" style operation:

(pseudocode):

```fsharp
PathProcessor = (File) -> boolean (GetNext | StopIteration)
```


```java
public void Iterator<File> contentOfDirectory(File directory, PathProcessor processor) throws IOException {
   try (val stream = Files.newDirectoryStream(directory.toPath())) {
      foreach (val path in stream) {
          if (!processor.process(path)) {
               break;
          }
      }
}
```

Arthur-Milchior commented:
I don't understand your pseudo code. in particular `(GetNext | StopIteration)` makes no sens to me

According to its implementation, Kotlin's `sequence` is just an extra layer for Iterator. It has a lot of really cool feature syntactically, it is far easier to create a sequence than an iterator, but it's really just an iterator. Any iterator trouble will occur also with sequence.

Also `foreach` is terminal. I.e. it is immediately applied to the entire sequence. Which entirely defeat my purpose to avoid processing tens of thousands of elements at once.

i thank you for calling my attention for the need to explicitly close the stream, I never noticed it. That seems indeed very counter-intuitive to me that some java object may be garbage collected and still hold ressources.

My intuition would be simply to call `close` when it cease to have content.
The cases i can imagine are:
* either everything goes right and it gets closed
* files are missing, but it also get closed
* the app crashes/is halted by the user/lack of battery, and the resources is not freed explicitly, but anyway the app do not retain resource when the OS deletes its process I assume.

So, while it's clearly not beautiful code, it really seems better than every other alternative that I can think of, at least as long as I want to ensure that each operation is executed very quickly

david-allison commented:
> I don't understand your pseudo code. in particular `(GetNext | StopIteration)` makes no sens to me

Too pseudo 😅. A boolean, where true represents "Process next element in the sequence" and false represents "stop iteration".

> Also `foreach` is terminal. I.e. it is immediately applied to the entire sequence. Which entirely defeat my purpose to avoid processing tens of thousands of elements at once.

I thought we were doing `stream.map(::toMoveOperation)` which is functionally equivalent to immediate application?

> i thank you for calling my attention for the need to explicitly close the stream, I never noticed it. That seems indeed very counter-intuitive to me that some java object may be garbage collected and still hold ressources.
> 
> My intuition would be simply to call `close` when it cease to have content. The cases i can imagine are:
> 
> * either everything goes right and it gets closed
> * files are missing, but it also get closed
> * the app crashes/is halted by the user/lack of battery, and the resources is not freed explicitly, but anyway the app do not retain resource when the OS deletes its process I assume.
> 
> So, while it's clearly not beautiful code, it really seems better than every other alternative that I can think of, at least as long as I want to ensure that each operation is executed very quickly

Would returning an `AutoCloseable` class wrapping the iterator work? Then we can use [.use { }](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/use.html) and it'll be closed without the need to worry about exception handling

App crashing is fine, JVM+OS will clear that up

david-allison commented on AnkiDroid/src/test/java/com/ichi2/compat/DirectoryContentTest.kt:
```diff
@@ -0,0 +1,84 @@
+                "two" -> 2
+                else -> -1
+            }
+            MatcherAssert.assertThat("File ${file.name} should not be in ${directory.path}", fileNumber, CoreMatchers.not(CoreMatchers.equalTo(-1)))
```
import `MatcherAssert` and `CoreMatchers`

david-allison commented on AnkiDroid/src/main/java/com/ichi2/compat/FileStream.kt:
```diff
@@ -0,0 +1,10 @@
+
+/**
+ * A type to read a sequence of [File]. It is essentially [DirectoryStream]<File>, but accessible
+ * in all API.
```
Note that it should explicitly be closed after use

david-allison reviewed (commented):
I like this. I'd approve once CI is green.

I would like to see a response to my comment on why `foreach` wouldn't be acceptable here, as that would further reduce complexity. But I'd be happy to see this go in as-is.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/compat/FileStream.kt:
```diff
@@ -0,0 +1,10 @@
+package com.ichi2.compat
```
missing copyright header (in case you missed the lint check)

Arthur-Milchior commented:
> I thought we were doing `stream.map(::toMoveOperation)` which is functionally equivalent to immediate application?

DirectoryStream is not actually a [Stream], so I can't use standard library stream methods such as map. But even if I were to do so, it would then be a mistake of my part. My goal is explicitly not to do any loop processing here if I can avoid it.
 
> Would returning an `AutoCloseable` class wrapping the iterator work? Then we can use [.use { }](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/use.html) and it'll be closed without the need to worry about exception handling

I'm sorry, but I don't understand. Do you want to use `use` only in test, as in your code comment (which would work but be of little interest). Or do you want to also use it in the actual code. I don't think I can use it in CompatV26. Indeed, `use` seems to take a function that will be entirely executed immediately. It does not seems that `use` can interact with a stream and wait until the stream is empty/no other method has a need of it anymore.

Arthur-Milchior commented on AnkiDroid/src/test/java/com/ichi2/compat/DirectoryContentTest.kt:
```diff
@@ -0,0 +1,84 @@
+        val directory = createTransientDirectory()
+        val iterator = CompatHelper.getCompat().contentOfDirectory(directory)!!
+        MatcherAssert.assertThat("Iterator should not have next", iterator.hasNext(), CoreMatchers.equalTo(false))
+        iterator.close()
```
I don't get it. Your suggestion seems actually less similar to what I'll actually use in my prod code. Indeed, the only time when i'll need this method instead of `listFiles` is when I'll actually need to stream, to proceed in small increment, which is when I can't use `use`.

i applied the suggestion, because essentially I'm also fine with it. Just not convinced it's an improvement

david-allison replied:
The new code is exception-safe.

david-allison commented:
> > I thought we were doing `stream.map(::toMoveOperation)` which is functionally equivalent to immediate application?
> 
> DirectoryStream is not actually a [Stream], so I can't use standard library stream methods such as map. But even if I were to do so, it would then be a mistake of my part. My goal is explicitly not to do any loop processing here if I can avoid it.

Do you have a PR explaining what you're trying to do as an end goal here? Or could you list how you're going to use this code.

I'm happy with the code as it is, but I don't understand why it's implemented the way it is, and the use case that we have that wouldn't expect the whole stream to be consumed.

> > Would returning an `AutoCloseable` class wrapping the iterator work? Then we can use [.use { }](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/use.html) and it'll be closed without the need to worry about exception handling
> 
> I'm sorry, but I don't understand. Do you want to use `use` only in test, as in your code comment (which would work but be of little interest). Or do you want to also use it in the actual code. I don't think I can use it in CompatV26. Indeed, `use` seems to take a function that will be entirely executed immediately. It does not seems that `use` can interact with a stream and wait until the stream is empty/no other method has a need of it anymore.

`use { }` ensures that the access is exception safe, it has nothing to do with streaming. It's a property of `AutoClosable`.

You want to use it on the value returned from `Compat`, not inside `Compat`.

Arthur-Milchior commented:
> Do you have a PR explaining what you're trying to do as an end goal here? Or could you list how you're going to use this code.
Here it is https://github.com/ankidroid/Anki-Android/pull/10364

> I'm happy with the code as it is, but I don't understand why it's implemented the way it is, and the use case that we have that wouldn't expect the whole stream to be consumed.

I expect the whole stream to be eventually consumed. My issue is that the whole stream is consumed all at once, which means that the thread was blocked until the whole stream was processed. With this change, I can consume the stream one element at a time, which ensure that each function returns very quickly.

> `use { }` ensures that the access is exception safe, it has nothing to do with streaming. It's a property of `AutoClosable`.

According to the documentation of `use`, it execute the block and then close the AutoCloseable element. Which mean that if I call `contentOfDirectory(dir).use{some_code(it)}`, `some_code` is executed once and then the stream is closed. Which is really not what I want, because i want the stream to remain open as long as the directory contains file and subdirectories that have not been processed. My very goal is to ensure that we never have to loop over the entire content of the directory, but that each time we ask the stream for the next file, it is received in a separate Operation.

Arthur-Milchior commented:
I modified the code to ensure that `hasNext` do not throw and just return `false` in case of error

david-allison commented:
> I modified the code to ensure that `hasNext` do not throw and just return `false` in case of error

This means the operation is reported as successful, when in reality the iterator failed

Arthur-Milchior commented:
I don't know what we can do if the operation failed, but here it is, I rewrote it so that it catches. I added an extra Exception to encapsulate `DirectoryIteratorException`, since this is not available in API < 26

david-allison commented on AnkiDroid/src/main/java/com/ichi2/compat/DirectoryIteratorException.kt:
```diff
@@ -0,0 +1,41 @@
+ * Encapsulate java.nio.file.DirectoryIteratorException
+ * and make it accessible to API < 26. *
+ */
+class DirectoryIteratorException(val exception: RuntimeException) : RuntimeException(exception) {
```
This should inherit from IOException and be documented.

Arthur-Milchior replied:
Which kind of documentation do you want?

david-allison replied:
After further reading of the Java `DirectoryIteratorException`, I think we can get rid of this class completely and unwrap the `IOException`.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/compat/FileStream.kt:
```diff
@@ -0,0 +1,27 @@
+ * If `hasNext` returned true, method `next` should not throw.
+ * @see [DirectoryStream]
+ */
+interface FileStream : Iterator<File>, AutoCloseable
```
Can we override `hasNext()` and add the exception via `@Throws()`?

david-allison replied:
This should work

```
    @Throws(IOException::class)
    override fun hasNext(): Boolean
```

david-allison commented on AnkiDroid/src/test/java/com/ichi2/compat/DirectoryContentTest.kt:
```diff
@@ -0,0 +1,95 @@
+
+    @Test
+    fun ensure_absolute_path() {
+        // Relative paths caused me hours of debugging. Never again.
+        val directory = createTransientDirectory()
+            .withTempFile("zero")
+        val iterator = CompatHelper.getCompat().contentOfDirectory(directory)!!
+        val file = iterator.next()
+        assertThat("Paths should be absolute", file.path, equalTo(file.absolutePath))
```
I think we want `canonicalPath` here. Canonical paths are absolute, absolute paths may not be canonical

Arthur-Milchior replied:
I'm fine with it, but it was clearly not what I was looking for.
Indeed, my trouble was that one of the path was local, which led to error when i was trying to access the file, as it was resolved in ankidroid/AnkiDroid.
Absolute path totally avoid this problem, even if not canonical. But indeed, no trouble in requiring canonical, I doubt in AnkiDroid we would ever get an error here

david-allison replied:
Implementer's choice. Happy you've considered it.

david-allison commented on AnkiDroid/src/main/java/com/ichi2/compat/DirectoryIteratorException.kt:
```diff
@@ -0,0 +1,41 @@
+ */
+class DirectoryIteratorException(val exception: RuntimeException) : RuntimeException(exception) {
+    init {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
```
Why are we throwing in the constructor of an exception?

Arthur-Milchior replied:
This is supposed to encapsulate a `DirectoryIteratorException`. The trouble is that I can't require the constructor to only accept `DirectoryIteratorException`, because otherwise the exception would only be available starting at API 26. Which mean that I could not try to catch this exception specifically in code that accept API 21.

This path is never supposed to be taken, however, this plus documentation is the most straightforward way I found to indicate that the argument must be a `DirectoryIteratorException`

Arthur-Milchior replied:
This is supposed to encapsulate a `DirectoryIteratorException`. The trouble is that I can't require the constructor to only accept `DirectoryIteratorException`, because otherwise the exception would only be available starting at API 26. Which mean that I could not try to catch this exception specifically in code that accept API 21.

This path is never supposed to be taken, however, this plus documentation is the most straightforward way I found to indicate that the argument must be a `DirectoryIteratorException`

david-allison replied:
Unless we specifically need this exception, I'd prefer if we unwrapped the `IOException` from the `DirectoryIteratorException` and threw it in `hasNext()`

To answer your question, we could add a [static factory method](https://stackoverflow.com/a/929273)

david-allison reviewed (commented):
I don't see much benefit in wrapping the exception. If we are going to go down this route, I'd prefer a checked exception coming from `hasNext()` if this is possible.

Do note that Kotlin doesn't have the concept of checked Exceptions. `@Throws` is used for compatibility with Java.

Arthur-Milchior commented:
> I don't see much benefit in wrapping the exception.

How do you want to deal with it otherwise? I can't catch the original `DirectoryIteratorException` in usual code, because this exception is only defined at API26 and higher. So the only thing I can do to still throw this exception and catch it in MoveDirectory is to encapsulate it in a class I create.

> I'd rather just have hasNext() throw, as it previously did.

This is what this code currently do. It captures the Stream's exception and encapsulate it in my exception. This way, in API up to 25 we can catch my exception (that can never be thrown actually), and in API 26 and up, we can catch my exception and access the inner exception if we want to analyze it.

> If we are going to go down this route, I'd prefer a checked exception coming from hasNext() if this is possible.
If I want `FileStream` to be an Iterator, then `hasNext` should not throw catched exception, the type system would not allow it. I can decide that I don't actually want it to be an iterator, after all, I'm probably not going to ever loop on its result, is it what you want?

david-allison commented on AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.java:
```diff
@@ -183,4 +186,46 @@ public void abandonAudioFocus(AudioManager audioManager, AudioManager.OnAudioFoc
+                    // According to the documentation, it's the only exception it can throws.
+                    // We encapsulate `e` in a class that is available in
+                    // API < 26.
+                    throw new com.ichi2.compat.DirectoryIteratorException(e);
```
Overall, I think removing `DirectoryIteratorException` would make the code a /lot/ simpler

```suggestion
                    throw e.getCause()
```

david-allison reviewed (changes requested):
On further review of the underlying Java `DirectoryIteratorException`, it feels like our `DirectoryIteratorException` isn't adding value. 

Would you have an issue with unwrapping the `IOException` in `CompatV26` removing our `DirectoryIteratorException`?

Arthur-Milchior commented:
Done. Hopefully

Arthur-Milchior commented:
I believe that fb30e0cc55fe15ca12c07c758abcaa3bcbecfbf9 is what you asked for. I'll remove FileSTream to be iterable

david-allison reviewed (approved):
Looks good to me, thanks!

Arthur-Milchior commented:
I just removed an outdated comment.

>                     // We encapsulate `e` in a class that is available in
                    // API < 26.


Appart from that, the code didn't change since David approved it

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/compat/CompatV26.java:
```diff
@@ -183,4 +186,44 @@ public void abandonAudioFocus(AudioManager audioManager, AudioManager.OnAudioFoc
+
+
+            @Override
+            public boolean hasNext() throws IOException{
```
interesting - how did this spacing thing slip through ktlint :thinking: ?

```suggestion
            public boolean hasNext() throws IOException {
```

Arthur-Milchior replied:
Because it's a java file

mikehardy replied:
oh of course :facepalm:

mikehardy commented on AnkiDroid/src/main/java/com/ichi2/compat/FileStream.kt:
```diff
@@ -0,0 +1,43 @@
+    fun hasNext(): Boolean
+
+    /**
+     * Guaranteed not to throw if [hasNext] returned true.
```
I checked out the docs since this seemed suspicious

They say "This means that if hasNext method returns true, and is followed by a call to the next method, then it is guaranteed that the next method will not fail with a DirectoryIteratorException." which is - I think - how you have coded it up here with regard to documentation about what throws and when.

But we are not in control of usage, and users may call next without calling hasNext. They should not, but they may.

I think this is easy to fix though, the full snippet of docs is:

> If an I/O error is encountered when accessing the directory then it causes the Iterator's hasNext or next methods to throw [DirectoryIteratorException](https://developer.android.com/reference/java/nio/file/DirectoryIteratorException) with the [IOException](https://developer.android.com/reference/java/io/IOException) as the cause. As stated above, the hasNext method is guaranteed to read-ahead by at least one element. This means that if hasNext method returns true, and is followed by a call to the next method, then it is guaranteed that the next method will not fail with a DirectoryIteratorException.

So I propose that this is marked with a @throws DirectoryIteratorException (or IOException if you catch DirectoryIteratorException - whatever seems best) and make the documentation *really explicit* that hasNext should always be called first. 

Really, that second bit - the documentation being much more firm about always calling hasNext first so it reads ahead one entry and thus makes calling next safe - is the important bit to me. Just to give developers the firm guidance that the documentation seems to be saying upstream

mikehardy reviewed (changes requested):
Everything looks really good I just had one nitpick I tried to communicate in a comment

Arthur-Milchior commented:
I hope this is what you expected. If the documentation do not please yo, I'd gladly ask you to rewrite it and let me know how you want it to be. 

And to answer your question, the missing space was in a java file, which should make clear why Kotlin tool did not correct it

Arthur-Milchior commented:
> But we are not in control of usage, and users may call next without calling hasNext. They should not, but they may.

I guess I see your point. Especially as I suggested to eventually move this as a separate library if anybody else have a usage for a preemptable move directory operation. But still, it's quite strange to read "we are not in control of usage" while it seems impossible for me to imagine that some part of scoped storage gets merged in without any of us three to review it, and quite improbable that this is used again

mikehardy reviewed (approved):
I think that's great - you also have a good point that okay, we are in control of it since we are reviewing things that merge, but as a favor to future developers and reviewers here, 10 years from now, the documentation is now exceptionally clear on the expectation to call hasNext first. Totally works for me, thanks Arthur

mikehardy commented:
#5304
